iree/compiler/Conversion/CodegenUtils/GPUUtils.h

```diff
@@ -0,0 +1,48 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//===----------------------------------------------------------------------===//
+//
+// Utility functions common to all GPU backends.
+//
+//===----------------------------------------------------------------------===//
+#ifndef IREE_COMPILER_CONVERSION_CODEGENUTILS_GPUUTILS_H_
+#define IREE_COMPILER_CONVERSION_CODEGENUTILS_GPUUTILS_H_
+
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Value.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/FoldUtils.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+static constexpr int kNumGPUDims = 3;
+
+/// Generate the operations that compute the processor ID and number of
+/// processors. Used as the callback needed for LinalgDistributionOptions.
+class GPUGlobalId;
+class GPUGlobalCount;
+template <typename GPUIdOp, typename GPUCountOp>
+SmallVector<linalg::ProcInfo, 2> getGPUProcessorIdsAndCounts(OpBuilder &builder,
```

Comment:
Actually having a templated API entry (since its in a header file) is problematic (very easy to have undefined symbols). Probably just better to keep them local to files.

Reply:
You're right, thinking more about this, it is definitely an overkill. I reverted all this and just made a local function with just what I need for CUDA.